File: baekjoon/dynamic_programming/11722_가장_긴_감소하는_부분_수열.cpp
#include <cstdio>

int n, ans = 0, arr[1001], dp[1001];

int getMax(int a, int b) { return a < b ? b : a; }

int main(void) {
	scanf("%d", &n);

	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
		dp[i] = 1;
	}

	for (int i = n - 1; i >= 0; i--) {
		for (int j = n - 1; j > i; j--) {
			if (arr[j] < arr[i] && dp[j] + 1 > dp[i]) {
				dp[i] = dp[j] + 1;
			}
		}
		ans = getMax(ans, dp[i]);
	}

	printf("%d", ans);

	return 0;
}

File: baekjoon/dynamic_programming/11722_가장긴감소하는부분수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
  public static void main(String[] args) throws IOException {
    int ans = 0;
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(br.readLine());
    String[] a = br.readLine().split(" ");
    int[] arr = new int[n];
    int[] dp = new int[n];

    for (int i = 0; i < n; i++) {
      arr[i] = Integer.parseInt(a[i]);
    }

    for (int i = 0; i < n; i++) {
      dp[i] = 1;
      for (int j = 0; j < i; j++) {
        if (arr[i] < arr[j] && dp[i] < dp[j] + 1) {
          dp[i] = dp[j] + 1;
        }
      }

      ans = Math.max(ans, dp[i]);
    }

    System.out.println(ans);
  }
}
